### PR TITLE
[FSW] Fix resource starvation by using a helper class

### PR DESCRIPTION
**Pull request by Therzok**
This PR should be reviewed commit by commit.

The idea is that it uses a helper tree that mimicks the real filesystem tree to prune the maximum amount of directories being watched.

### CODE REVIEW

**Comment by Therzok** [comment]
Huh, will take a look at the tests

**Comment by mrward** [comment]
I think the problem is that the old WatchDirectories used to always remove null or empty directories. Now if a HashSet is passed to WatchDirectories this does not happen. Workbench passes a HashSet to the WatchDirectories and this can include null or empty paths for a document even though the Document indicates it is a file. I guess maybe we should have a test for that.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs [hunk 1]
```diff
@@ -26,180 +26,181 @@
...
+				return modified;
+
+			// Apply new ones if we have any
+			var set = new HashSet<FilePath> (directories.Where(x => !x.IsNullOrEmpty));
+
+			if (set.Count > 0) {
+				monitoredDirectories [id] = set;
+				foreach (var path in set) {
+					if (path.IsNullOrEmpty)
+						continue;
```
Comment (mrward) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: Will this line be reached? It looks to me that the set is created each time and IsNullOrEmpty items are removed before this so the continue will not be called.

**Comment by Therzok** [comment]
With a build from this PR, I don't see the IDE running out of handles when going in idle.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs (hunk 1 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects/FileWatcherService.cs]: Left-over. Will remove.

**Comment by Therzok** [comment]
Some FSW is not unregistered and causing deadlocks on shutdown. will fix.

**Comment by Therzok** [comment]
@slluis PTAL, should be solid now.

**Comment by Therzok** [comment]
@monojenkins backport release-7.6

**Pull request merged**